src/qbe_ir_generator.cpp

```diff
@@ -144,65 +176,115 @@ void QbeIrGenerator::Visit(const DeclStmtNode& decl_stmt) {
-        WriteInstr_("storel {}, {}", FuncScopeTemp{init_num},
+        WriteInstr_("storew {}, {}", FuncScopeTemp{init_num},
                     FuncScopeTemp{id_num});
       }
-    } else {
-      WriteInstr_("storew {}, {}", FuncScopeTemp{init_num},
-                  FuncScopeTemp{id_num});
     }
+    // Set up the number of the id so we know were to load it back.
+    id_to_num[decl.id] = id_num;
   }
-  // Set up the number of the id so we know were to load it back.
-  id_to_num[decl.id] = id_num;
 }
 
 void QbeIrGenerator::Visit(const ArrDeclNode& arr_decl) {
-  int base_addr_num = NextLocalNum();
-  assert(arr_decl.type->IsArr());
-  const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
-  auto element_size = arr_type->element_type().size();
-  WriteInstr_("{} =l alloc{} {}", FuncScopeTemp{base_addr_num}, element_size,
-              arr_decl.type->size());
-  id_to_num[arr_decl.id] = base_addr_num;
-
-  for (auto i = std::size_t{0}, e = arr_type->len(); i < e; ++i) {
-    if (i < arr_decl.init_list.size()) {
+  if (arr_decl.is_global) {
+    const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
+    assert(arr_type);
+    Write_("export data {} = align {} {{ ",
+           user_defined::GlobalPointer{arr_decl.id},
+           arr_type->element_type().size());
+
+    global_var_init_vals.clear();
+    auto arr_size = arr_type->len();
+    auto init_len = arr_decl.init_list.size();
+    // The predicate of this loop guarantees that it doesn't go out of bound if
+    // the number of initialized elements is less than the array declaration
+    // size.
+    for (auto i = std::size_t{0}; i < arr_size && i < init_len; ++i) {
       auto& arr_init = arr_decl.init_list.at(i);
       arr_init->Accept(*this);
+      Write_("{}", GenerateQBEInit(global_var_init_vals.at(i)));
+      if (i != arr_size - 1) {
+        Write_(", ");
+      }
     }
 
-    const int offset = NextLocalNum();
-    WriteInstr_("{} =l extsw {}", FuncScopeTemp{offset}, i * element_size);
+    // set remaining elements as 0
+    if (init_len < arr_size) {
+      Write_("z {}", (arr_size - init_len) * arr_type->element_type().size());
+    }
+    Write_(" }}\n");
+  } else {
+    int base_addr_num = NextLocalNum();
+    assert(arr_decl.type->IsArr());
+    const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
+    auto element_size = arr_type->element_type().size();
+    WriteInstr_("{} =l alloc{} {}", FuncScopeTemp{base_addr_num}, element_size,
+                arr_decl.type->size());
+    id_to_num[arr_decl.id] = base_addr_num;
+
+    // NOTE: Compiler will not set elements to 0 if `init_len` is 0.
+    // 6.7.9 Initialization
+    // 10. If an object that has automatic storage duration is not initialized
+    // explicitly, its value is indeterminate.
+    for (auto i = std::size_t{0}, e = arr_type->len(),
+              init_len = arr_decl.init_list.size();
+         i < e && 0 < init_len; ++i) {
```

Comment:
Since `init_len` is an unsigned, use `!= 0` to better convey the condition that it should have at least one init value.
```suggestion
         i < e && init_len != 0; ++i) {
```